src/peft/tuners/hira/model.py

```diff
@@ -0,0 +1,761 @@
+        >>> model = prepare_model_for_kbit_training(model)
+        >>> hira_model = get_peft_model(model, config)
+        ```
+
+    **Attributes**:
+        - **model** ([`~transformers.PreTrainedModel`]) -- The model to be adapted.
+        - **peft_config** ([`LoraConfig`]): The configuration of the Lora model.
+    """
+
+    prefix: str = "hira_"
+
+    def __init__(
+        self,
+        model,
+        config,
+        adapter_name,
+        low_cpu_mem_usage: bool = False,
+        state_dict: Optional[dict[str, torch.Tensor]] = None,
+    ) -> None:
+        super().__init__(
+            model,
+            config,
+            adapter_name,
+            low_cpu_mem_usage=low_cpu_mem_usage,
+            state_dict=state_dict,
+        )
+
+    def _check_new_adapter_config(self, config: HiRAConfig) -> None:
+        """
+        A helper method to check the config when a new adapter is being added.
+
+        Raise a ValueError if there is something wrong with the config or if it conflicts with existing adapters.
+
+        """
+        return True
+
+    @staticmethod
+    def _check_target_module_exists(hira_config, key):
+        return check_target_module_exists(hira_config, key)
+
+    def _prepare_model(self, peft_config: HiRAConfig, model: nn.Module):
+        r"""
+        A private method to modify the model structure before adapter is applied.
+
+        Args:
+            peft_config (`PeftConfig`):
+                The prepared adapter config.
+            model (`nn.Module`):
+                The model that is going to be adapted.
+        """
+        if peft_config.layer_replication:
+            replicate_layers(model, peft_config.layer_replication)
+
+    def _create_and_replace(
+        self,
+        hira_config,
+        adapter_name,
+        target,
+        target_name,
+        parent,
+        current_key,
+    ):
+        if current_key is None:
+            raise ValueError("Current Key shouldn't be `None`")
+
+        # Regexp matching - Find key which matches current target_name in patterns provided
+        r_key = get_pattern_key(hira_config.r_pattern.keys(), current_key)
+        r = hira_config.r_pattern.get(r_key, hira_config.r)
+
+        kwargs = {
+            "r": r,
+            "hira_dropout": hira_config.hira_dropout,
+            "fan_in_fan_out": hira_config.fan_in_fan_out,
+            "init_hira_weights": hira_config.init_hira_weights,
+            "loaded_in_8bit": getattr(self.model, "is_loaded_in_8bit", False),
+            "loaded_in_4bit": getattr(self.model, "is_loaded_in_4bit", False),
+        }
+        # for torchao merging, we need the get_apply_tensor_subclass from the quantization config
+        try:
+            kwargs["get_apply_tensor_subclass"] = operator.attrgetter(
+                "hf_quantizer.quantization_config.get_apply_tensor_subclass"
+            )(self.model)
+        except AttributeError:
+            pass
+
+        quant_methods = ["gptq", "aqlm", "awq"]
```

Comment:
Only bnb is supported, right? Let's remove the code dealing with other quant methods.